quark/subproject.py

```diff
@@ -705,6 +713,287 @@ def set_local_ignores(self, subprojects_dir, modules):
+                pipeline["id"],
+            ),
+            headers={
+                self.gitlab_token_header: self.gitlab_token,
+            },
+        )
+
+        with urllib.request.urlopen(req) as response:
+            jobs = json.loads(response.read().decode("utf-8"))
+
+        for job in jobs:
+            if job["name"] == job_name:
+                return str(job["id"])
+
+        raise QuarkError("Could not find job %s in pipeline %s" % (job_name, pipeline["id"]))
+
+    def _download(self, tempdir: str) -> str:
+        assert self.parsed_artifact_name  # Make Pyright happy
+        assert self.parsed_endpoint_url  # Make Pyright happy
+
+        print_msg("downloading " + self.parsed_artifact_name, self._comment())
+
+        dl_dir = join(tempdir, "dl")
+        archive_path = join(dl_dir, self.parsed_artifact_name)
+
+        os.mkdir(dl_dir)
+
+        req = urllib.request.Request(
+            url="%s/api/v4/%s" % (self.gitlab_url, self.parsed_endpoint_url.lstrip("/")),
+            headers={
+                self.gitlab_token_header: self.gitlab_token,
+            },
+        )
+
+        with urllib.request.urlopen(req) as response:
+            with open(archive_path, "wb") as fileobj:
+                shutil.copyfileobj(response, fileobj)
+
+        # Verify or update the SHA1
+        print_msg("verifying " + self.parsed_artifact_name, self._comment())
+        with open(archive_path, "rb") as fileobj:
+            sha1 = hashlib.sha1(fileobj.read()).hexdigest()
+
+        if self.stamp["sha1"] and sha1 != self.stamp["sha1"]:
+            raise QuarkError("SHA1 mismatch: %s != %s" % (sha1, self.stamp["sha1"]))
+
+        self.stamp["sha1"] = sha1
+
+        return archive_path
+
+    def _extract(self, archive_path: str, tempdir: str):
+        assert self.directory  # Make Pyright happy
+
+        print_msg("extracting " + self.parsed_artifact_name, self._comment())
+
+        extract_dir = join(tempdir, "extract")
+        os.mkdir(extract_dir)
+
+        if archive_path.endswith((".tar.bz2", ".tar.gz", ".tar.xz")):
+            with tarfile.open(archive_path, "r") as tar:
+                tar.extractall(extract_dir)
+        elif archive_path.endswith(".zip"):
+            with zipfile.ZipFile(archive_path, "r") as zip:
+                zip.extractall(extract_dir)
+        else:
+            raise QuarkError("Unsupported format: " + archive_path)
+
+        # Move files into place
+        os.mkdir(self.directory)
+        items = os.listdir(extract_dir)
+        if len(items) == 1 and isdir(join(extract_dir, items[0])):
+            # Archive created a single directory, move its contents
+            for i in os.listdir(join(extract_dir, items[0])):
+                shutil.move(join(extract_dir, items[0], i), self.directory)
+        else:
+            # Archive is a tarbomb
+            for i in items:
+                shutil.move(join(extract_dir, i), self.directory)
```

Comment:
This too is a convenience that is nice for a user-facing tool, but having different behavior depending on the fact that there's a single directory or multiple entries in the archive is not ok... that means that suddenly an artifact that was extracted in a directory can be under a subdirectory if the target project decides to provide multiple files.